### PR TITLE
Ssh

### PR DESCRIPTION
**Pull request by lgund**
A lot of Bugfixes like in pull request #57. But now it's also possible to create ssh connections and Telnet connections. See example here #58 

I would like to prefer this as new mayor version 2.0. So we could have 1 big break and then we don't need to talk about this anymore 👍

### CODE REVIEW

File: src/TeamSpeak3QueryApi/Enums/TeamspeakConnectionType.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace TeamSpeak3QueryApi.Net.Enums
+{
+    public enum TeamspeakConnectionType
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Enums/TeamspeakConnectionType.cs:7]: I think something with "Protocol" would be a better fit.

File: src/TeamSpeak3QueryApi/FileTransfer/FileTransferClient.cs [hunk 2]
```diff
@@ -20,43 +20,43 @@ public int GetFileTransferId()
...
             using var ns = client.GetStream();
 
-            // Send key
+            // SendAsync key
             var keyBytes = Encoding.ASCII.GetBytes(key);
             await ns.WriteAsync(keyBytes, 0, keyBytes.Length).ConfigureAwait(false);
             await ns.FlushAsync().ConfigureAwait(false);
 
-            // Send data
+            // SendAsync data
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/FileTransfer/FileTransferClient.cs:49]: Acceidental renames?

File: src/TeamSpeak3QueryApi/Interfaces/ITypeCaster.cs [hunk 3]
```diff
@@ -2,8 +2,9 @@
 using System.Collections.Generic;
 using System.Globalization;
 using System.Linq;
+using TeamSpeak3QueryApi.Net.Extensions;
 
-namespace TeamSpeak3QueryApi.Net.Specialized
+namespace TeamSpeak3QueryApi.Net.Interfaces
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Interfaces/ITypeCaster.cs:7]: I think "interfaces" is not a good name for a namespace. We should come up with something more descriptive.

File: src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-namespace TeamSpeak3QueryApi.Net.Specialized.Notifications
+namespace TeamSpeak3QueryApi.Net.Notifications
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs:1]: Why was this moved?

File: src/TeamSpeak3QueryApi/Notifications/ClientMoved.cs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-namespace TeamSpeak3QueryApi.Net.Specialized.Notifications
+namespace TeamSpeak3QueryApi.Net.Notifications
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Notifications/ClientMoved.cs:1]: The `Specialized` contained everything related to the rich client (not the query client). That separation made sense to me.

File: src/TeamSpeak3QueryApi/Parameters/Parameter.cs [hunk 6]
```diff
@@ -1,7 +1,7 @@
 using System;
 using System.Diagnostics;
 
-namespace TeamSpeak3QueryApi.Net
+namespace TeamSpeak3QueryApi.Net.Parameters
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Parameters/Parameter.cs:4]: The parameters are part of the query client and therefore belong to the upper namespace.

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs [hunk 7]
```diff
@@ -39,22 +55,26 @@ public class QueryClient : IDisposable
         private readonly Queue<QueryCommand> _queue = new Queue<QueryCommand>();
         private readonly ConcurrentDictionary<string, List<Action<NotificationData>>> _subscriptions = new ConcurrentDictionary<string, List<Action<NotificationData>>>();
 
+        SshClient _sshClient;
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/QueryClient.cs:58]: I think a better abstraction would be creating derived classses of a common base class.

This would also prevent having the wrong default port for different protocols.

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs [hunk 8]
```diff
@@ -82,19 +106,52 @@ public async Task<CancellationTokenSource> Connect()
...
+            {
+                throw new QueryProtocolException("Telnet Query isn't a valid Teamspeak Query");
+            }
             await _reader.ReadLineAsync().ConfigureAwait(false); // Ignore welcome message
             await _reader.ReadLineAsync().ConfigureAwait(false);
 
             return ResponseProcessingLoop();
         }
 
+        public CancellationTokenSource ConnectSsh(string username, string password)
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/QueryClient.cs:123]: Doesn't TS support public key auth via SSH?

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs [hunk 9]
```diff
@@ -350,25 +409,32 @@ private CancellationTokenSource ResponseProcessingLoop()
                     string line = null;
                     try
                     {
-                        line = await _reader.ReadLineAsync().WithCancellation(cts.Token).ConfigureAwait(false);
+                        if(ConnectionType == TeamspeakConnectionType.SSH)
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/QueryClient.cs:412]: This should be handled in a base class of a common sub class for TCP and SSH.

File: src/TeamSpeak3QueryApi/Query/QueryCommand.cs [hunk 10]
```diff
@@ -1,7 +1,8 @@
 using System.Collections.Generic;
 using System.Threading.Tasks;
+using TeamSpeak3QueryApi.Net.Parameters;
 
-namespace TeamSpeak3QueryApi.Net
+namespace TeamSpeak3QueryApi.Net.Query
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/QueryCommand.cs:5]: This is part of the query client and belongs to the `TeamSpeak3QueryApi.Net` namespace.

File: src/TeamSpeak3QueryApi/Responses/ClientBan.cs [hunk 11]
```diff
@@ -1,4 +1,4 @@
-namespace TeamSpeak3QueryApi.Net.Specialized.Responses
+namespace TeamSpeak3QueryApi.Net.Responses
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Responses/ClientBan.cs:1]: Parsed responses of a rich client do not belong to the same namespace as the query client.

File: src/TeamSpeak3QueryApi/Enums/GetClientOptions.cs [hunk 12]
```diff
@@ -1,6 +1,6 @@
 using System;
 
-namespace TeamSpeak3QueryApi.Net.Specialized
+namespace TeamSpeak3QueryApi.Net.Enums
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Enums/GetClientOptions.cs:3]: Same for this as the other comments.

File: src/TeamSpeak3QueryApi.Demo/Program.cs [hunk 13]
```diff
@@ -19,29 +19,34 @@ static async Task Main(string[] args)
             var user = loginData[1].Trim();
             var password = loginData[2].Trim();
 
-            var rc = new TeamSpeakClient(host);
+            // Ssh connection
+            var rc = new TeamSpeakClient(host, TeamspeakConnectionType.SSH);
+            rc.ConnectSsh(user, password);
 
-            await rc.Connect();
+            // Telnet connection
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi.Demo/Program.cs:26]: The readme needs to be updated, too.

**Comment by nikeee** [comment]
Thanks for the PR!
I reviewed parts of it. It's such a large PR that I cannot review this entirely in my spare time.
I think the best way to approach the review is to read all comments and then do some changes.

File: src/TeamSpeak3QueryApi/Interfaces/ITypeCaster.cs (hunk 3 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Interfaces/ITypeCaster.cs:7]: I couldn't change this to just interface cause we have a interface class... Any ideas?

File: src/TeamSpeak3QueryApi/FileTransfer/FileTransferClient.cs (hunk 2 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/FileTransfer/FileTransferClient.cs:49]: I think yes. Every async method should be changed. When we have breaking changes then we shouldn't make some quick and dirty solutions

File: src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs (hunk 4 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs:1]: I've removed the specialized folder cause i couldn't do anything with it

File: src/TeamSpeak3QueryApi/Notifications/ClientMoved.cs (hunk 5 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Notifications/ClientMoved.cs:1]: Ah okey. But doesn't it make more sense to put the query connection to separate folder? So we have the primary class which will be used in the root folder

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs (hunk 7 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/QueryClient.cs:58]: Confirmed. But we should do this as separate pr?

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs (hunk 8 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/QueryClient.cs:123]: First of all teamspeak support key Auth. But I am not sure if there is a public key Auth. Need investigation

**Comment by lgund** [comment]
All changes they have a like will be changed until next week. I am this week not @home.

I know this is a big pr but this isn't a final version for sure. I just want to create a final version with ssh support for everyone. I am sure this brench need some investigation.

File: src/TeamSpeak3QueryApi/Query/QueryCommand.cs (hunk 10 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/QueryCommand.cs:5]: Cause a query command is still part of the query client. So it should stay in the query folder.

**Comment by lgund** [comment]
So I´am done. Let me know if something else should be fixed or changed.

File: src/TeamSpeak3QueryApi/Query/Protocols/IProtocol.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace TeamSpeak3QueryApi.Net.Query.Protocols
+{
+    public interface IProtocol
+    {
+        Task<CancellationTokenSource> ConnectAsync();
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/Protocols/IProtocol.cs:11]: That interface doesn't make any sense IMHO when only either of these Methods can be implemented.

File: src/TeamSpeak3QueryApi/Query/Protocols/IProtocol.cs [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+        CancellationTokenSource Connect(string username, string password);
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/Protocols/IProtocol.cs:12]: A `CancellationToken` should be returned, not a `CancellationTokenSource`

File: src/TeamSpeak3QueryApi/Query/Protocols/SshProtocol.cs [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Renci.SshNet;
+using Renci.SshNet.Common;
+using TeamSpeak3QueryApi.Net.Enums;
+using TeamSpeak3QueryApi.Net.Extensions;
+
+namespace TeamSpeak3QueryApi.Net.Query.Protocols
+{
+    public class SshProtocol : QueryClient, IProtocol
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/Protocols/SshProtocol.cs:16]: I'd name this `SshQueryClient` and the other one `TelnetQueryClient` or `RawQueryClient` (which is technically more correct).

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs [hunk 17]
```diff
@@ -31,128 +30,52 @@ public class QueryClient : IDisposable
...
 
-            return ResponseProcessingLoop();
+            OnDisconnected?.Invoke(this, EventArgs.Empty);
+            _cts.Cancel();
         }
 
-        public CancellationTokenSource ConnectSsh(string username, string password)
+        /// <summary>Connects to the Query API server.</summary>
+        /// <returns>An awaitable <see cref="Task"/>.</returns>
+        public virtual Task<CancellationTokenSource> ConnectAsync()
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/QueryClient.cs:69]: This should be abstract or non-existent (it will be there in the child classes).

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs [hunk 18]
```diff
@@ -31,128 +30,52 @@ public class QueryClient : IDisposable
 
         /// <summary>Gets the remote host of the Query API client.</summary>
         /// <returns>The remote host of the Query API client.</returns>
-        public string Host { get; }
+        public string Host { get; internal set; }
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/QueryClient.cs:33]: Why internal?

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs [hunk 19]
```diff
@@ -31,128 +30,52 @@ public class QueryClient : IDisposable
...
+        /// <summary>Connects to the Query API server.</summary>
+        /// <returns>An awaitable <see cref="Task"/>.</returns>
+        public virtual CancellationTokenSource Connect(string username, string password)
         {
-            if (_cts == null)
-                return;
-
-            OnDisconnected?.Invoke(this, EventArgs.Empty);
-            _cts.Cancel();
+            throw new NotImplementedException();
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Query/QueryClient.cs:78]: Same as at `public virtual Task<CancellationTokenSource> ConnectAsync()`

File: src/TeamSpeak3QueryApi/Query/QueryCommand.cs (hunk 10 above)
Comment (nikeee) [reply src/TeamSpeak3QueryApi/Query/QueryCommand.cs:5]: The namespcae layout is as follows:

`namespace TeamSpeak3QueryApi.Net`: Everything related to the Query protocol itself (that also the regular TS client uses)
`namespace TeamSpeak3QueryApi.Net.Specialised`: While the name `Specialised` is disputable, this contains the concrete server query implementation.

One core principle is that the child namespaces should only rely on the parent namespaces (not some sibling namespace) if possible. Introducing a `Query` namespace would break this semantic.

File: src/TeamSpeak3QueryApi/TeamSpeakClient.cs [hunk 20]
```diff
@@ -15,6 +15,7 @@
 using TeamSpeak3QueryApi.Net.Query.Parameters;
 using TeamSpeak3QueryApi.Net.Query.Responses;
 using TeamSpeak3QueryApi.Net.Query.Enums;
+using TeamSpeak3QueryApi.Net.Query.Protocols;
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/TeamSpeakClient.cs:18]: While I think namespaces like `Enums` etc. might be good for _applications_, I don't think that a library should use that pattern. For example `System.Windows.Forms` doesn't have something like `Controls`, `Components` etc. either. This way, the core principle mentioned above is easier to follow.

**Review by nikeee — CHANGES_REQUESTED** [review]

File: src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs (hunk 4 above)
Comment (nikeee) [reply src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs:1]: It is a concrete implementation of the notification that the server sends and therefore belongs to the `Specialized` namespace.

File: src/TeamSpeak3QueryApi/Notifications/ClientMoved.cs (hunk 5 above)
Comment (nikeee) [reply src/TeamSpeak3QueryApi/Notifications/ClientMoved.cs:1]: See other comments on this.

File: src/TeamSpeak3QueryApi/Enums/Protocol.cs [hunk 21]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace TeamSpeak3QueryApi.Net.Enums
+{
+    public enum Protocol
```
Comment (nikeee) [inline src/TeamSpeak3QueryApi/Enums/Protocol.cs:7]: What is the enum used for?

```
QueryClient c0 = new SshQueryClient();
QueryClient c1 = new RawQueryClient();
```
...would not need such an enum.

File: src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs (hunk 4 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Notifications/ChannelDeleted.cs:1]: Okey sure... the notifications is part of the query and should be moved there

File: src/TeamSpeak3QueryApi/Query/QueryCommand.cs (hunk 10 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/QueryCommand.cs:5]: You mean I should change the class name QueryClient to just Client cause it´s in the namespace Query? (For example)

File: src/TeamSpeak3QueryApi/Enums/Protocol.cs (hunk 21 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Enums/Protocol.cs:7]: The interface was needed cause the query client hasn´t the methods. Now was a abstract declaration we don´t need this interface anymore. Thanks for this information 👍

File: src/TeamSpeak3QueryApi/Query/Protocols/SshProtocol.cs (hunk 16 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/Protocols/SshProtocol.cs:16]: Confirmed. It´s much better of it´s extends from a query client.

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs (hunk 17 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/QueryClient.cs:69]: Changed to abstract.

File: src/TeamSpeak3QueryApi/Query/QueryClient.cs (hunk 18 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/QueryClient.cs:33]: I don´t want that somebody outside this class could change the host. But the Ssh or Telnet client need to set the host and for that it must get more access.

File: src/TeamSpeak3QueryApi/Query/Protocols/IProtocol.cs (hunk 15 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/Query/Protocols/IProtocol.cs:12]: Why? This function is not changed. You also returned the `CancellationTokenSource`. It´s needed for the processingloop. Did I miss something?

File: src/TeamSpeak3QueryApi/TeamSpeakClient.cs (hunk 20 above)
Comment (lgund) [reply src/TeamSpeak3QueryApi/TeamSpeakClient.cs:18]: I think this is a matter of taste. I would like to stay like this. It´s for me better readable if we stay in folder and not change the classnames. Or we could change the classnames and stay also in the folders?

**Comment by lgund** [comment]
Is still interested in this pr, if not I would release it as a new nuget package.

**Comment by nikeee** [comment]
I'm still interested in merging this! :)

Currently, I'm short on spare time for this project. Maybe we can settle this down this weekend.
